File: lib/src/main/java/com/orientsec/easysocket/utils/NetUtils.java
package com.orientsec.easysocket.utils;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.os.Build;

/**
 * Utility class `NetUtils` provides methods for network-related operations.
 */
public class NetUtils {

    /**
     * Checks whether the network is available.
     *
     * @param context The context used to access system services.
     * @return `true` if the network is available, `false` otherwise.
     */
    public static boolean isNetworkAvailable(Context context) {
        // Get the ConnectivityManager system service
        ConnectivityManager cm =
                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (cm == null) return false;

        // Check network status for devices running below Android M
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            NetworkInfo info = cm.getActiveNetworkInfo();
            return info != null && info.isConnected();
        } else {
            // Check network capabilities for devices running Android M and above
            NetworkCapabilities nc = cm.getNetworkCapabilities(cm.getActiveNetwork());
            return nc != null && nc.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED);
        }
    }
}

File: library/core/src/androidMain/kotlin/com/orientsec/easysocket/utils/NetUtils.kt
package com.orientsec.easysocket.utils

import android.Manifest
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import androidx.annotation.RequiresPermission

/**
 * Utility class `NetUtils` provides methods for network-related operations.
 */
object NetUtils {

    /**
     * Checks whether the network is available.
     *
     * @param context The context used to access system services.
     * @return `true` if the network is available, `false` otherwise.
     */
    @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
    @JvmStatic
    fun isNetworkAvailable(context: Context?): Boolean {
        if (context == null) return false
        // Get the ConnectivityManager system service
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return false

        // Check network status for devices running below Android M
        // Check network capabilities for devices running Android M and above
        val nc = cm.getNetworkCapabilities(cm.activeNetwork)
        return nc != null && nc.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
    }
}
